classes/controllers/tab/workflow/PKPVersioningTabHandler.inc.php

```diff
@@ -0,0 +1,204 @@
+ * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+ *
+ * @class ReviewRoundTabHandler
+ * @ingroup controllers_tab_workflow
+ *
+ * @brief Handle AJAX operations for version tabs on production stages workflow pages.
+ */
+
+// Import the base Handler.
+import('classes.handler.Handler');
+import('lib.pkp.classes.core.JSONMessage');
+
+class PKPVersioningTabHandler extends Handler {
+
+	/**
+	 * Constructor
+	 */
+	function __construct() {
+		parent::__construct();
+	}
+
+
+	//
+	// Extended methods from Handler
+	//
+	/**
+	 * @copydoc PKPHandler::authorize()
+	 */
+	function authorize($request, &$args, $roleAssignments) {
+		// We need a submission revision id in request.
+		import('lib.pkp.classes.security.authorization.internal.VersioningRequiredPolicy');
+		$this->addPolicy(new VersioningRequiredPolicy($request, $args));
+
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+
+	/**
+	 * create new submission revision
+	 * @param $args array
+	 * @param $request PKPRequest
+	 * @return JSONMessage JSON object
+	 */
+	function newVersion($args, $request){
+		$submissionId = (int)$request->getUserVar('submissionId');
+		$submissionDao = Application::getSubmissionDAO();
+		$submissionRevision = $submissionDao->getLatestRevisionId($submissionId);
+		// get data of the old version
+		$submission = $submissionDao->getById($submissionId, null, false, $submissionRevision);
+		// authors
+		$authorDao = DAORegistry::getDAO('AuthorDAO');
+		$authors = $authorDao->getBySubmissionId($submissionId, true, false, $submissionRevision);
+		// galleys
+		import('classes.article.ArticleGalley');
+		$articleGalleyDao = DAORegistry::getDAO('ArticleGalleyDAO');
+		$articleGalleys = $articleGalleyDao->getBySubmissionId($submissionId, null, $submissionRevision);
+		$galleys = $articleGalleys->toArray();
+
+		// update submission version and remove publication date
+		$submissionRevision++;
+		$submission->setData('submissionRevision', $submissionRevision);
+		$submission->setDatePublished(null);
+
+		// save new submission version
+		$submissionDao->updateObject($submission);
+
+		// copy the authors from old version to new version
+		foreach($authors as $author) {
+			$authorId = (int)$author->getId();
+			$author->setVersion($submissionRevision);
+			$authorDao->insertObject($author, true);
+			$newAuthorId = (int)$authorDao->getInsertId();
+			$authorDao->update('UPDATE authors SET author_id = ? WHERE author_id = ?', array($authorId, $newAuthorId));
+		}
+
+		// create new galley versions
+		foreach($galleys as $galley) {
+			// copy file and link copy to new galley
+			if($galley->getFile()){
+				$context = $request->getContext();
+				$oldFile = $galley->getFile();
+				$fileStage = $oldFile->getFileStage();
+				$newFileId = $this->copyFile($context, $oldFile, $fileStage);
+				$galley->setFileId($newFileId);
+			}
+			// save new galley version
+			$articleGalleyDao->updateObject($galley);
```

Comment:
hmmm... why do we update the whole object? could we not only add new files relations?